### ISSUE TITLE
Warning when using transactionStop

### ISSUE THREAD

**Issue by idoadiv** [PommeVerte/gremlin-php#17]
Hi,

After calling transactionStop I get a warning in my gremlin server log "WARN  org.apache.tinkerpop.gremlin.server.handler.OpSelectorHandler  - Message with op code [close] is not recognized".

PHP Code:

$db->transactionStart();
$db->send($query)
$db->transactionStop(true);

Full trace:
[gremlin-server-worker-1] WARN  org.apache.tinkerpop.gremlin.server.handler.OpSelectorHandler  - Message with op code [close] is not recognized.
org.apache.tinkerpop.gremlin.server.op.OpProcessorException: Message with op code [close] is not recognized.
	at org.apache.tinkerpop.gremlin.server.op.AbstractEvalOpProcessor.select(AbstractEvalOpProcessor.java:100)
	at org.apache.tinkerpop.gremlin.server.handler.OpSelectorHandler.decode(OpSelectorHandler.java:80)
	at org.apache.tinkerpop.gremlin.server.handler.OpSelectorHandler.decode(OpSelectorHandler.java:50)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.http.websocketx.WebSocketServerProtocolHandler$1.channelRead(WebSocketServerProtocolHandler.java:146)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:182)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:130)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
	at java.lang.Thread.run(Thread.java:745)

**Comment by PommeVerte** [comment]
Yes this is normal. `close` was introduced in the upcoming gremlin-server 3.1.0. I added the feature early as normally appart from the warning this shouldn't have any incidence on your workflow. Let me know if it does.

3.1.0 should hit the shelves sometime next week (it's currently in VOTE). I encourage you to upgrade as it corrects quite a few outstanding session issues.

**Comment by PommeVerte** [comment]
I'm going to keep this open as a reminder to create a test covering proper session closing... I thought I had one but can't find it. Maybe it got lost in some branch I never merged.

**Comment by PommeVerte** [comment]
Info so I don't forget:

There actually is a multichannel test case for this: https://github.com/PommeVerte/gremlin-php/blob/master/tests/RexsterTest.php#L236-L279

I need to add a single channel test as well.

**Comment by PommeVerte** [comment]
Ok Thanks for reporting this. Here's the status:

The warning you're seeing is because of a piece of code I set up to correct an issue with sessions not closing properly on `$db->close()`. This is a server side bug that affected `gremlin-server 3.0.2` and is corrected for `3.1.0`.

The [test I linked to earlier](https://github.com/PommeVerte/gremlin-php/blob/master/tests/RexsterTest.php#L236-L279) was a test to see if the session had properly been closed on `$db->close()`. This test **should** be failing for `3.0.2` yet it passes and sessions are closed properly (why?). 

This had me a little baffled until I realized that the code I setup early to support `gremlin-server 3.1.0` session close features actually generated an exception (the warning you reported) that resulted in the terminating of the session. It's incredibly hacky but the warning you see actually means that the server side bug was forcibly corrected by the `gremlin-php` driver. 

So I guess the warning is a good thing.

I mentioned that this was tested in a multi-channel context and that I needed to add single channel tests, but there's no point in doing this as there's no single-channel use case for this.

Cheers.

So all in all I've added a couple of lines to existing tests "just in case" and will be closing this issue.
